File: backend/app/engine/tour.py
"""Multi-hop tour planner: find transfer options and chain them."""

from dataclasses import dataclass
from typing import Callable

import astropy.units as u
from astropy.time import Time

from app.engine.bodies import PLANETS
from app.engine.cache import cache_date_range
from app.engine.launch import LaunchWindow, find_next_window

# on_progress(depth, origin, destination, local_index, local_total)
ProgressCallback = Callable[[int, str, str, int, int], None]


@dataclass(frozen=True)
class TourOption:
    window: LaunchWindow
    wait_time: u.Quantity  # time waiting before launch
    next_options: tuple  # tuple[TourOption, ...], empty if leaf


@dataclass(frozen=True)
class TourNode:
    origin: str
    start_date: Time
    options: tuple  # tuple[TourOption, ...]


def _count_windows(depth: int, n_planets: int = len(PLANETS)) -> int:
    """Total number of window searches across all depths.

    At each level, we exclude the current planet, giving (n_planets - 1) destinations.
    Each destination then has (n_planets - 1) sub-destinations (excluding itself).
    """
    n_dest = n_planets - 1
    if depth <= 1:
        return n_dest
    return n_dest + n_dest * _count_windows(depth - 1, n_planets)


def plan_tour(origin: str, start_date: Time, depth: int = 2,
              on_progress: ProgressCallback | None = None) -> TourNode:
    """Plan a multi-hop tour from origin at start_date.

    Args:
        origin: Starting planet name.
        start_date: Earliest departure date.
        depth: How many hops to compute (1 = direct only, 2 = one relay, etc.).
        on_progress: Optional callback(depth, origin, dest, local_i, local_total).

    Returns:
        TourNode with options for each reachable destination, each containing
        nested next_options if depth > 1.
    """
    start_iso = start_date.iso[:10]
    cache_start, cache_end = cache_date_range()
    if start_iso >= cache_end:
        raise ValueError(
            f"Start date {start_iso} is beyond the cache range. "
            f"Cache covers {cache_start} to {cache_end}."
        )

    max_depth = depth
    options = _find_options(origin, start_date, depth, max_depth, on_progress)

    if not options:
        raise ValueError(
            f"No launch windows found for {origin} after {start_iso}. "
            f"Cache covers {cache_start} to {cache_end}."
        )

    return TourNode(
        origin=options[0].window.origin if options else origin,
        start_date=start_date,
        options=tuple(options),
    )


def _find_options(origin: str, after: Time, depth: int, max_depth: int,
                  on_progress: ProgressCallback | None) -> list[TourOption]:
    """Find all transfer options from origin after a given date."""
    destinations = [p for p in PLANETS if p.name.lower() != origin.lower()]
    hop_level = max_depth - depth + 1
    options = []
    for i, planet in enumerate(destinations, 1):
        try:
            window = find_next_window(origin, planet.name, after)
        except ValueError:
            continue  # no cached window available for this pair/date
        wait_time = (window.launch_date - after).to(u.day)

        if on_progress:
            on_progress(hop_level, origin.capitalize(), planet.name, i, len(destinations))

        if depth > 1:
            next_opts = _find_options(
                window.destination, window.arrival_date, depth - 1,
                max_depth, on_progress,
            )
        else:
            next_opts = []

        options.append(TourOption(
            window=window,
            wait_time=wait_time,
            next_options=tuple(next_opts),
        ))
    return options

File: frontend/tour.js
// ─── Client-side tour planner ───
// Port of backend/app/engine/tour.py, using cache data directly.

import { lookupWindow, cacheDateRange, getPlanets } from './cache.js';

/**
 * Add days to an ISO date string, return new ISO date string.
 * Uses millisecond arithmetic to match Python's TimeDelta behavior.
 */
function addDays(isoDate, days) {
  const ms = new Date(isoDate + 'T00:00:00Z').getTime() + days * 86400000;
  return new Date(ms).toISOString().slice(0, 10);
}

/**
 * Compute days between two ISO date strings.
 */
function daysBetween(isoA, isoB) {
  const a = new Date(isoA + 'T12:00:00Z');
  const b = new Date(isoB + 'T12:00:00Z');
  return (b - a) / 86400000;
}

/**
 * Plan a multi-hop tour from origin at startDateIso.
 *
 * @param {string} origin - Starting planet name
 * @param {string} startDateIso - Earliest departure date (YYYY-MM-DD)
 * @param {number} depth - How many hops (1 = direct, 2 = one relay, etc.)
 * @returns {object} TourResponse-shaped object with origin, start_date, options[]
 */
export function planTour(origin, startDateIso, depth) {
  const [, cacheEnd] = cacheDateRange();
  if (startDateIso >= cacheEnd) {
    throw new Error(
      `Start date ${startDateIso} is beyond the cache range. Cache ends ${cacheEnd}.`
    );
  }

  const planets = getPlanets();
  const options = findOptions(origin, startDateIso, depth, planets);

  if (!options.length) {
    const [cacheStart] = cacheDateRange();
    throw new Error(
      `No launch windows found for ${origin} after ${startDateIso}. ` +
      `Cache covers ${cacheStart} to ${cacheEnd}.`
    );
  }

  return {
    origin: options[0].window.origin,
    start_date: startDateIso,
    options,
  };
}

function findOptions(origin, afterIso, depth, planets) {
  const destinations = planets.filter(
    p => p.name.toLowerCase() !== origin.toLowerCase()
  );

  const options = [];
  for (const planet of destinations) {
    const entry = lookupWindow(origin, planet.name, afterIso);
    if (!entry) continue;

    const launchDate = entry.launch;
    const transferDays = entry.transfer_time_days;
    const arrivalDate = addDays(launchDate, transferDays);
    const waitDays = daysBetween(afterIso, launchDate);

    const window = {
      origin: origin.charAt(0).toUpperCase() + origin.slice(1).toLowerCase(),
      destination: planet.name,
      launch_date: launchDate,
      arrival_date: arrivalDate,
      transfer_time_days: transferDays,
      departure_dv_km_s: entry.departure_dv_km_s,
      arrival_dv_km_s: entry.arrival_dv_km_s,
      delta_v_total_km_s: entry.delta_v_total_km_s,
    };

    let nextOptions = [];
    if (depth > 1) {
      nextOptions = findOptions(planet.name, arrivalDate, depth - 1, planets);
    }

    options.push({
      window,
      wait_time_days: Math.round(waitDays * 10000) / 10000,
      next_options: nextOptions,
    });
  }

  return options;
}
